### PR TITLE
[WIP] : Update loop inside validatePayment to loop over proving periods (Issue #252)

### PR DESCRIPTION
**Pull request by parzival1821**
## Related Issue
Closes 252 ( Once completed)

## Current Progress
I've implemented the updated looping algorithm in validatePayment as described in the issue
- [x] Replaced the existing loop with new algorithm 
- [ ] Convert provenPeriods to a bitmap
- [ ] Remove provenThisPeriod

## What I Need Feedback On
**Before I continue with the rest of the issue, I'd like to get feedback on my implementation of the looping algorithm.**

Since there were no existing tests for this specific part of the code, I want to make sure my approach is correct before implementing the remaining parts of the issue. Also I would love to help you guys write some tests for the same if required

## Questions for Maintainers
- Does the new looping logic match what you had in mind?
- Are there any edge cases I should consider?
- Any suggestions for improvement before I move forward?

## Changes Made
- Modified `FilecoinWarmStorageService.sol` to implement the new loop structure

## Next Steps (after approval)
Once you confirm this approach is correct, I'll:
- Complete the remaining tasks in #252
- Request a full review

**Note:** This is a WIP/Draft PR specifically for early feedback on the algorithm implementation.
(Also, just a heads-up, I changed my username from `dev-n-dough` to `parzival1821`

### CODE REVIEW

**Comment by parzival1821** [comment]
@wjmelements @rjan90 Would love to get your thoughts on this draft PR

**Comment by parzival1821** [comment]
Some workflow checks are failing, sorry about that, let me just fix them real quick

File: service_contracts/foundry.lock [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  "lib/openzeppelin-contracts-upgradeable": {
+    "rev": "3bfc52f2fbf5be95de632f346169dac6a669bd57"
+  },
+  "lib/pdp": {
+    "rev": "61681392933926fbccb142ab7767e037680850b4"
+  },
+  "lib/session-key-registry": {
+    "rev": "e472ca2b525fb2396832216182b64a0c165cb49c"
+  }
+}
```
Comment (wjmelements) [inline service_contracts/foundry.lock:20]: remove

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 2]
```diff
@@ -1582,6 +1574,69 @@ contract FilecoinWarmStorageService is
         });
     }
 
+    function findProvenEpochs(uint256 dataSetId, uint256 fromEpoch, uint256 toEpoch)
+        public
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol:1578]: internal

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 3]
```diff
@@ -1582,6 +1574,69 @@ contract FilecoinWarmStorageService is
...
+                    lastProvenEpoch = toEpoch;
+                }
+            }
+        }
+
+        return (provenEpochCount, lastProvenEpoch);
+    }
+
+    function _isPeriodProven(uint256 dataSetId, uint256 periodId) private view returns (bool) {
+        uint256 currentPeriod = getProvingPeriodForEpoch(dataSetId, block.number);
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol:1629]: only determine currentPeriod once

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 4]
```diff
@@ -1547,15 +1547,7 @@ contract FilecoinWarmStorageService is
         uint256 provenEpochCount = 0;
         uint256 lastProvenEpoch = fromEpoch;
 
-        // Check each epoch in the range
-        for (uint256 epoch = fromEpoch + 1; epoch <= toEpoch; epoch++) {
-            bool isProven = isEpochProven(dataSetId, epoch);
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol:1552]: can we remove isEpochProven?

**Review by wjmelements — CHANGES_REQUESTED** [review]

**Comment by parzival1821** [comment]
Yes I think something went wrong with the rebasing, I am trying to fix that

**Comment by rjan90** [comment]
@parzival1821 Some smaller Lint Check failures to fix up here

File: .gitignore [hunk 5]
```diff
@@ -1,2 +1,4 @@
 # VIM
 *.swp
+foundry.lock
```
[the remark this answers is not in the document]
Comment (wjmelements) [reply .gitignore]: actually this probably only belongs in the `service_contracts` dir

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 6]
```diff
@@ -1545,6 +1500,72 @@ contract FilecoinWarmStorageService is
...
     }
 
+    function _findProvenEpochs(uint256 dataSetId, uint256 fromEpoch, uint256 toEpoch)
+        internal
+        view
+        returns (uint256, uint256)
+    {
+        uint256 provenEpochCount = 0;
+        uint256 lastProvenEpoch = fromEpoch;
+        uint256 activationEpoch = provingActivationEpoch[dataSetId];
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]: reuse this `sload` from where it was first accessed when checking `"Proving never activated for this data set"`

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 7]
```diff
@@ -1545,6 +1500,72 @@ contract FilecoinWarmStorageService is
...
+        uint256 currentPeriod = getProvingPeriodForEpoch(dataSetId, block.number);
+        if (toEpoch >= activationEpoch && toEpoch < block.number) {
```
[the remark this answers is not in the document]
Comment (wjmelements) [reply service_contracts/src/FilecoinWarmStorageService.sol]: the first pattern scales like this
```
if (conditionA) {
  if (conditionB) {
    if (conditionC) {
        /*many lines of code*/
        return happyD;
    } else {
      return nullC;
    }
  } else {
    return nullB;
  }
} else {
  return nullA;
}
```
the second pattern scales like this
```
if (!conditionA) {
   return nullA;
}
if (!conditionB) {
  return nullB;
}
if (!conditionC) {
  return nullC;
}
/*many lines of code*/
return happyD;
```
`require` assertions follow this second pattern.

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 8]
```diff
@@ -1545,6 +1500,72 @@ contract FilecoinWarmStorageService is
...
+            // if `toEpoch` lies after activation, and `fromEpoch` lies before activation, then update the `fromEpoch`, as follows :
+            if (fromEpoch < activationEpoch - 1) {
+                fromEpoch = activationEpoch - 1; 
+            }
+
+            uint256 startingPeriod = getProvingPeriodForEpoch(dataSetId, fromEpoch + 1);
+            uint256 endingPeriod = getProvingPeriodForEpoch(dataSetId, toEpoch);
+
+            // lets handle first period separately
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]:
```suggestion
            // handle first period separately
```

File: service_contracts/src/FilecoinWarmStorageService.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (wjmelements) [reply service_contracts/src/FilecoinWarmStorageService.sol]: Also update the documentation of that error to include this situation.

**Comment by wjmelements** [comment]
I think your core logic is correct. I will look at it again after you update to reason about it some more. I am most worried about off by one errors.

File: service_contracts/src/FilecoinWarmStorageService.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (parzival1821) [reply service_contracts/src/FilecoinWarmStorageService.sol]: Got it. Same error must be emitted in the case of `toEpoch < activationEpoch` right?

File: service_contracts/lib/forge-std
[the remark this answers is not in the document]
Comment (wjmelements) [reply service_contracts/lib/forge-std]: please

File: service_contracts/src/Errors.sol [hunk 9]
```diff
@@ -201,7 +201,8 @@ library Errors {
     /// @param railId The rail ID
     error RailNotAssociated(uint256 railId);
 
-    /// @notice The epoch range is invalid (toEpoch must be > fromEpoch)
+    /// @notice The epoch range is invalid
+    /// @notice Will be emitted if any of the following conditions is NOT met - (fromEpoch must be less than toEpoch), (toEpoch must be greater than block number), (toEpoch must be greater than the activation epoch)
```
Comment (wjmelements) [inline service_contracts/src/Errors.sol]: list these on separate lines, without parentheses

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 10]
```diff
@@ -1545,6 +1500,73 @@ contract FilecoinWarmStorageService is
...
+    function _isPeriodProven(uint256 dataSetId, uint256 periodId, uint256 currentPeriod) private view returns (bool) {
+        if (periodId == currentPeriod) {
+            return provenThisPeriod[dataSetId];
+        }
+        uint256 isProven = provenPeriods[dataSetId][periodId >> 8] & (1 << (periodId & 255));
+        return isProven != 0;
+    }
+
+    function _calcPeriodDeadline(uint256 dataSetId, uint256 periodId) private view returns (uint256) {
+        return provingActivationEpoch[dataSetId] + ((periodId + 1) * maxProvingPeriod); // we need to do `periodId + 1` since it starts from 0
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]:
```suggestion
        return provingActivationEpoch[dataSetId] + (periodId + 1) * maxProvingPeriod;
```

File: service_contracts/lib/forge-std
[the remark this answers is not in the document]
Comment (wjmelements) [reply service_contracts/lib/forge-std]: I have pushed a revert, ffc358ef2ed9d0a4ae5347f4c673a8552b28bd5a

File: service_contracts/src/Errors.sol [hunk 11]
```diff
@@ -201,7 +201,11 @@ library Errors {
     /// @param railId The rail ID
     error RailNotAssociated(uint256 railId);
 
-    /// @notice The epoch range is invalid (toEpoch must be > fromEpoch)
+    /// @notice The epoch range is invalid
+    /// @notice Will be emitted if any of the following conditions is NOT met :
+    /// @notice 1. fromEpoch must be less than toEpoch
+    /// @notice 2. toEpoch must be greater than block number
```
Comment (wjmelements) [inline service_contracts/src/Errors.sol]: in payments, toEpoch can be equal to `block.number`.

File: service_contracts/src/FilecoinWarmStorageService.sol (hunk 6 above)
Comment (wjmelements) [reply service_contracts/src/FilecoinWarmStorageService.sol]: please

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 12]
```diff
@@ -1545,6 +1501,75 @@ contract FilecoinWarmStorageService is
...
+        uint256 activationEpoch = provingActivationEpoch[dataSetId];
+        if (toEpoch < activationEpoch) {
+            revert Errors.InvalidEpochRange(fromEpoch, toEpoch);
+        }
+        if (toEpoch > block.number) {
+            revert Errors.InvalidEpochRange(fromEpoch, toEpoch);
+        }
+        uint256 currentPeriod = getProvingPeriodForEpoch(dataSetId, block.number);
+
+        // if `toEpoch` lies after activation, and `fromEpoch` lies before activation, then update the `fromEpoch`, as follows :
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]: you should not have spaces before your colons

Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]: remove this comment

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 13]
```diff
@@ -1545,6 +1501,74 @@ contract FilecoinWarmStorageService is
...
+            for (uint256 period = startingPeriod + 1; period < endingPeriod; period++) {
+                if (_isPeriodProven(dataSetId, period, currentPeriod)) {
+                    provenEpochCount += maxProvingPeriod;
+                    lastProvenEpoch = _calcPeriodDeadline(dataSetId, period);
+                }
+            }
+
+            // now handle the last period separately
+            if (_isPeriodProven(dataSetId, endingPeriod, currentPeriod)) {
+                // then the epochs to add = `endingPeriodStarting` to `toEpoch`. But since `endingPeriodStarting` is simply the ending of its previous period + 1, so  epochs : (`deadline + 1` -> `to`)
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]: do not resolve comments when they are not resolved
https://github.com/FilOzone/filecoin-services/pull/267#discussion_r2445718379

File: service_contracts/src/FilecoinWarmStorageService.sol (hunk 6 above)
Comment (wjmelements) [reply service_contracts/src/FilecoinWarmStorageService.sol]: > Actually activationEpoch is not used earlier near line 1464,

Yes it is

Screenshot 2025-10-20 at 2 50 48 PM


> Do you still want me to move this variable to top level validatePayment function and then pass it down to _findProvenEpochs fn?

Yes I do

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 14]
```diff
@@ -1494,11 +1457,12 @@ contract FilecoinWarmStorageService is
         require(dataSetId != 0, Errors.RailNotAssociated(railId));
 
         // Calculate the total number of epochs in the requested range
-        uint256 totalEpochsRequested = toEpoch - fromEpoch;
+        uint256 totalEpochsRequested = toEpoch - fromEpoch; // total epochs = (toEpoch) - (fromEpoch + 1) - 1 == toEpoch - fromEpoch
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]: don't need this comment

**Comment by wjmelements** [comment]
> I think your core logic is correct. I will look at it again after you update to reason about it some more. I am most worried about off by one errors.

I have re-reviewed the logic. It is correct because the start epoch is exclusive and the end epoch is inclusive.

**Comment by wjmelements** [comment]
> So this PR is ready to be merged sir? @wjmelements

no
https://github.com/FilOzone/filecoin-services/pull/267/files#r2445854987

File: service_contracts/lib/openzeppelin-contracts
Comment (wjmelements) [inline service_contracts/lib/openzeppelin-contracts]: you still haven't reverted your changes to the submodules

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 15]
```diff
@@ -1514,6 +1471,74 @@ contract FilecoinWarmStorageService is
...
+        uint256 provenEpochCount,
+        uint256 lastProvenEpoch,
+        uint256 activationEpoch
+    ) internal view returns (uint256, uint256) {
+        if (toEpoch < activationEpoch) {
+            revert Errors.InvalidEpochRange(fromEpoch, toEpoch);
+        }
+        if (toEpoch > block.number) {
+            revert Errors.InvalidEpochRange(fromEpoch, toEpoch);
+        }
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]:
```suggestion
        require (toEpoch >= activationEpoch && toEpoch <= block.number, Errors.InvalidEpochRange(fromEpoch, toEpoch));
```

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 16]
```diff
@@ -1530,8 +1521,8 @@ contract FilecoinWarmStorageService is
         return isProven != 0;
     }
 
-    function _calcPeriodDeadline(uint256 dataSetId, uint256 periodId) private view returns (uint256) {
-        return provingActivationEpoch[dataSetId] + (periodId + 1) * maxProvingPeriod;
```
Comment (wjmelements) [inline service_contracts/src/FilecoinWarmStorageService.sol]: I pass the activationEpoch from the parent function instead of looking it up again with the dataSetId

File: service_contracts/src/FilecoinWarmStorageService.sol [hunk 17]
```diff
@@ -1471,55 +1468,49 @@ contract FilecoinWarmStorageService is
         });
     }
 
-    function _findProvenEpochs(
-        uint256 dataSetId,
-        uint256 fromEpoch,
-        uint256 toEpoch,
-        uint256 provenEpochCount,
-        uint256 lastProvenEpoch,
```
[the remark this answers is not in the document]
Comment (wjmelements) [reply service_contracts/src/FilecoinWarmStorageService.sol]: #321 
c448c413c6824255cc773eb33994ea73366228a6

**Comment by parzival1821** [comment]
Thank you very much @wjmelements @rjan90 for guiding me through this PR. Would love to keep contributing to FilOzone and the Filecoin ecosystem as a whole!
